controllers/imagerepository_controller.go

```diff
@@ -49,190 +44,251 @@ import (
 }
 
-type dockerConfig struct {
-	Auths map[string]authn.AuthConfig
+func (r *ImageRepositoryReconciler) SetupWithManager(mgr ctrl.Manager, opts ImageRepositoryReconcilerOptions) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&imagev1.ImageRepository{}).
+		WithEventFilter(predicate.Or(predicate.GenerationChangedPredicate{}, predicates.ReconcileRequestedPredicate{})).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: opts.MaxConcurrentReconciles,
+			RateLimiter:             opts.RateLimiter,
+			RecoverPanic:            true,
+		}).
+		Complete(r)
 }
 
-// +kubebuilder:rbac:groups=image.toolkit.fluxcd.io,resources=imagerepositories,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups=image.toolkit.fluxcd.io,resources=imagerepositories/status,verbs=get;update;patch
-// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
-// +kubebuilder:rbac:groups="",resources=events,verbs=create;patch
-// +kubebuilder:rbac:groups="",resources=serviceaccounts,verbs=get;list;watch
-func (r *ImageRepositoryReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	reconcileStart := time.Now()
-
-	// NB: In general, if an error is returned then controller-runtime
-	// will requeue the request with back-off. In the following this
-	// is usually made explicit by _also_ returning
-	// `ctrl.Result{Requeue: true}`.
+func (r *ImageRepositoryReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, retErr error) {
+	start := time.Now()
+	log := ctrl.LoggerFrom(ctx)
 
-	var imageRepo imagev1.ImageRepository
-	if err := r.Get(ctx, req.NamespacedName, &imageRepo); err != nil {
+	// Fetch the ImageRepository.
+	obj := &imagev1.ImageRepository{}
+	if err := r.Get(ctx, req.NamespacedName, obj); err != nil {
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	defer r.recordSuspension(ctx, imageRepo)
-
-	log := ctrl.LoggerFrom(ctx)
+	// Record suspended status metric.
+	r.RecordSuspend(ctx, obj, obj.Spec.Suspend)
 
-	// Add our finalizer if it does not exist.
-	if !controllerutil.ContainsFinalizer(&imageRepo, imagev1.ImageRepositoryFinalizer) {
-		patch := client.MergeFrom(imageRepo.DeepCopy())
-		controllerutil.AddFinalizer(&imageRepo, imagev1.ImageRepositoryFinalizer)
-		if err := r.Patch(ctx, &imageRepo, patch); err != nil {
-			log.Error(err, "unable to register finalizer")
-			return ctrl.Result{}, err
-		}
+	// Return early if the object is suspended.
+	if obj.Spec.Suspend {
+		log.Info("reconciliation is suspended for this object")
+		return ctrl.Result{}, nil
 	}
 
-	// If the object is under deletion, record the readiness, and remove our finalizer.
-	if !imageRepo.ObjectMeta.DeletionTimestamp.IsZero() {
-		r.recordReadinessMetric(ctx, &imageRepo)
-		controllerutil.RemoveFinalizer(&imageRepo, imagev1.ImageRepositoryFinalizer)
-		if err := r.Update(ctx, &imageRepo); err != nil {
-			return ctrl.Result{}, err
```

Comment:
I think that moving the suspend check before deletion, makes all suspended resources to hand forever on deletion, since the controller no longer removes the finalizer. Also SC has the same bug after refactoring cc @hiddeco